docs/internals/filesystems.md

```diff
@@ -0,0 +1,138 @@
+and for fault tolerance and recovery. It is therefore crucial that the persistence semantics of these streams are well defined.
+
+## Definition of Persistence Guarantees
+
+Data written to an output stream is considered persistent, if two requirements are met:
+
+  1. **Visibility Requirement:** It must be guaranteed that all other processes, machines,
+     virtual machines, containers, etc. that are able to access the file see the data consistently
+     when given the absolute file path. This requirement is similar to the *close-to-open*
+     semantics defined by POSIX, but restricted to the file itself (by its absolute path).
+
+  2. **Durability Requirement:** The file system's specific durability/persistence requirements
+     must be met. These are specific to the particular file system. For example the
+     {@link LocalFileSystem} does not provide any durability guarantees for crashes of both
+     hardware and operating system, while replicated distributed file systems (like HDFS)
+     guarantee typically durability in the presence of up to concurrent failure or *n*
+     nodes, where *n* is the replication factor.
+
+Updates to the file's parent directory (such as that the file shows up when
+listing the directory contents) are not required to be complete for the data in the file stream
+to be considered persistent. This relaxation is important for file systems where updates to
+directory contents are only eventually consistent.
+
+The `FSDataOutputStream` has to guarantee data persistence for the written bytes once the call to
+`FSDataOutputStream.close()` returns.
+
+## Examples
+ 
+  - For **fault-tolerant distributed file systems**, data is considered persistent once 
+    it has been received and acknowledged by the file system, typically by having been replicated
+    to a quorum of machines (*durability requirement*). In addition the absolute file path
+    must be visible to all other machines that will potentially access the file (*visibility requirement*).
+
+    Whether data has hit non-volatile storage on the storage nodes depends on the specific
+    guarantees of the particular file system.
+
+    The metadata updates to the file's parent directory are not required to have reached
+    a consistent state. It is permissible that some machines see the file when listing the parent
+    directory's contents while other do not, as long as access to the file by its absolute path
+    is possible on all nodes.
+
+  - A **local file system** must support the POSIX *close-to-open* semantics.
+    Because the local file system does not have any fault tolerance guarantees, no further
+    requirements exist.
+ 
+    The above implies specifically that data may still be in the OS cache when considered
+    persistent from the local file system's perspective. Crashes that cause the OS cache loose
+    data are considered fatal to the local machine and not covered by the local file system's
+    guarantees as defined by Flink.
+
```

Comment:
Crashes that cause the OS cache to lose data are considered fatal to the local machine and are not covered by the local file system's guarantees as defined by Flink.